File: packages/core/test/pdf/pdfDocument.test.ts
import { unzlibSync } from 'fflate'
import { describe, expect, it } from 'vitest'
import { ArrayBufferSink, writePdf, type PdfDocument } from '../../src/pdf/pdfDocument.js'
import { PdfPage } from '../../src/pdf/pdfPage.js'
import { textWidth } from '../../src/pdf/metrics.js'

function write(build: (document: PdfDocument) => void): Uint8Array {
  const sink = new ArrayBufferSink()
  writePdf(sink, 595.276, 841.89, build)
  return sink.toBytes()
}

function latin1(bytes: Uint8Array): string {
  let s = ''
  for (const b of bytes) s += String.fromCharCode(b)
  return s
}

/** Content streams are deflated, so page operators have to be inflated to be read. */
function contentStreams(pdf: Uint8Array): string {
  const text = latin1(pdf)
  const header = /<< \/Filter \/FlateDecode \/Length (\d+) >>\nstream\n/g
  const parts: string[] = []
  let m: RegExpExecArray | null
  while ((m = header.exec(text)) !== null) {
    const start = m.index + m[0].length
    const length = Number(m[1])
    // unzlib, not inflate: a raw-deflate reader accepts both and would hide the bug that
    // shipped a PDF no viewer could open.
    parts.push(latin1(unzlibSync(pdf.subarray(start, start + length))))
  }
  return parts.join('\n')
}

describe('PdfDocument', () => {
  it('produces a structurally valid single page document', () => {
    const bytes = write((d) => d.addPage((page) => page.text(50, 50, 10, 'Page 1')))
    const text = latin1(bytes)
    expect(text.startsWith('%PDF-1.4')).toBe(true)
    expect(text.trimEnd().endsWith('%%EOF')).toBe(true)
    expect(text).toContain('/Type /Catalog')
    expect(text).toContain('/Type /Pages /Count 1')
    expect(text).toContain('/MediaBox [0 0 595.2760 841.8900]')
  })

  it('collects every page into one document', () => {
    const bytes = write((d) => {
      for (let i = 0; i < 9; i++) d.addPage((page) => page.text(10, 10, 8, `page ${i}`))
    })
    const text = latin1(bytes)
    expect(text).toContain('/Type /Pages /Count 9')
    expect(text.match(/\/Type \/Page \/Parent/g)).toHaveLength(9)
    const kids = text.split('/Kids [')[1]!.split(']')[0]!.trim().split(' 0 R').length - 1
    expect(kids).toBe(9)
  })

  it('every xref offset points at its own object header', () => {
    const bytes = write((d) => {
      for (let i = 0; i < 3; i++) {
        d.addPage((page) => {
          page.drawJpeg(Uint8Array.from({ length: 16 }, (_, k) => k), 4, 4, 0, 0, 10, 10)
          page.text(10, 10, 8, `page ${i}`)
        })
      }
    })
    const text = latin1(bytes)
    // "startxref" also ends in "xref", so anchor on the newline before the table.
    const xrefAt = text.lastIndexOf('\nxref\n') + 1
    const startXref = Number(text.split('startxref\n').pop()!.split('\n')[0]!.trim())
    expect(startXref).toBe(xrefAt)

    const lines = text.slice(xrefAt).split('\n')
    const count = Number(lines[1]!.trim().split(' ')[1])
    // Two fonts, the tree, the catalogue, then image + content + dictionary per page,
    // plus the mandatory free entry at number 0.
    expect(count).toBe(4 + 3 * 3 + 1)
    for (let index = 1; index < count; index++) {
      const offset = Number(lines[2 + index]!.slice(0, 10))
      expect(text.startsWith(`${index} 0 obj`, offset), `object ${index} at ${offset}`).toBe(true)
    }
    expect(text).toContain(`/Size ${count}`)
  })

  it('pages point at the page tree that is written after them', () => {
    const bytes = write((d) => {
      for (let i = 0; i < 2; i++) d.addPage((page) => page.text(0, 0, 8, 'x'))
    })
    const text = latin1(bytes)
    const parents = new Set([...text.matchAll(/\/Parent (\d+) 0 R/g)].map((m) => m[1]))
    expect(parents).toEqual(new Set(['3']))
    expect(text).toContain('3 0 obj\n<< /Type /Pages')
    expect(text).toContain('/Root 4 0 R')
  })

  it('jpeg payloads are embedded untouched as DCTDecode streams', () => {
    const jpeg = Uint8Array.from({ length: 64 }, (_, i) => (i * 7) & 0xff)
    const bytes = write((d) => d.addPage((page) => page.drawJpeg(jpeg, 8, 8, 0, 0, 100, 100)))
    const text = latin1(bytes)
    expect(text).toContain('/Subtype /Image /Width 8 /Height 8')
    expect(text).toContain('/Filter /DCTDecode /Length 64')
    expect(text).toContain(latin1(jpeg))
    expect(contentStreams(bytes)).toContain('100.0000 0 0 100.0000 0.0000 0.0000 cm /Im0 Do')
  })

  it('parentheses and backslashes in labels are escaped', () => {
    const bytes = write((d) => d.addPage((page) => page.text(0, 0, 8, 'a(b)c\\d')))
    expect(contentStreams(bytes)).toContain('(a\\(b\\)c\\\\d) Tj')
  })

  it('transcodes typographic punctuation to WinAnsi', () => {
    const bytes = write((d) => d.addPage((page) => page.text(0, 0, 8, '© IGN — SCAN25®')))
    const expected = `(© IGN ${String.fromCharCode(0x97)} SCAN25®) Tj`
    expect(contentStreams(bytes)).toContain(expected)
  })

  it('deflates content streams', () => {
    const bytes = write((d) =>
      d.addPage((page) => {
        for (let i = 0; i < 40; i++) page.text(10, i, 8, 'un libelle assez repetitif')
      }),
    )
    const compressed = Number(/<< \/Filter \/FlateDecode \/Length (\d+) >>/.exec(latin1(bytes))![1])
    const plain = contentStreams(bytes).length
    expect(compressed, `${compressed} vs ${plain}`).toBeLessThan(plain / 2)
  })

  it('uses the real Helvetica metrics for text width', () => {
    const page = new PdfPage(595.276, 841.89)
    // Every Helvetica digit is 556/1000 em.
    expect(page.textWidth('1234', 10)).toBeCloseTo(4 * 0.556 * 10, 9)
    expect(page.textWidth('iii', 10)).toBeLessThan(page.textWidth('WWW', 10))
    expect(textWidth('Page', 10, true)).toBeGreaterThan(textWidth('Page', 10))
  })

  it('refuses an empty document', () => {
    expect(() => write(() => {})).toThrow(/at least one page/)
  })
})

describe('content stream framing', () => {
  it('frames content streams as zlib, which is what /FlateDecode means', () => {
    // fflate's deflateSync emits raw RFC 1951. Poppler reports "Unknown compression method
    // in flate stream" and renders a blank page — which is exactly what shipped once.
    const bytes = write((d) => d.addPage((page) => page.text(50, 50, 10, 'Page 1')))
    const text = latin1(bytes)
    const header = /<< \/Filter \/FlateDecode \/Length (\d+) >>\nstream\n/.exec(text)
    expect(header).not.toBeNull()
    const start = header!.index + header![0].length
    const stream = bytes.subarray(start, start + Number(header![1]))
    // RFC 1950: low nibble 8 is DEFLATE, and the two header bytes are a multiple of 31.
    expect(stream[0]! & 0x0f).toBe(8)
    expect(((stream[0]! << 8) | stream[1]!) % 31).toBe(0)
    expect(latin1(unzlibSync(stream))).toContain('(Page 1) Tj')
  })
})

File: core/src/test/kotlin/io/gpxtoign/core/pdf/PdfDocumentTest.kt
package io.gpxtoign.core.pdf

import java.io.ByteArrayOutputStream
import java.util.zip.Inflater
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class PdfDocumentTest {

    private fun write(build: (PdfDocument) -> Unit): ByteArray {
        val out = ByteArrayOutputStream()
        writePdf(out, 595.276, 841.89, build)
        return out.toByteArray()
    }

    private fun latin1(bytes: ByteArray) = bytes.toString(Charsets.ISO_8859_1)

    /** Content streams are deflated, so page operators have to be inflated to be read. */
    private fun contentStreams(pdf: ByteArray): String {
        val text = latin1(pdf)
        val header = Regex("<< /Filter /FlateDecode /Length (\\d+) >>\nstream\n")
        return header.findAll(text).joinToString("\n") { match ->
            val start = match.range.last + 1
            val length = match.groupValues[1].toInt()
            val inflater = Inflater()
            inflater.setInput(pdf, start, length)
            val out = ByteArrayOutputStream()
            val buffer = ByteArray(8192)
            while (!inflater.finished()) {
                val n = inflater.inflate(buffer)
                if (n == 0) break
                out.write(buffer, 0, n)
            }
            inflater.end()
            latin1(out.toByteArray())
        }
    }

    @Test
    fun `produces a structurally valid single page document`() {
        val bytes = write { it.addPage { page -> page.text(50.0, 50.0, 10.0, "Page 1") } }
        val text = latin1(bytes)
        assertTrue(text.startsWith("%PDF-1.4"), "missing header")
        assertTrue(text.trimEnd().endsWith("%%EOF"), "missing trailer")
        assertTrue(text.contains("/Type /Catalog"))
        assertTrue(text.contains("/Type /Pages /Count 1"))
        assertTrue(text.contains("/MediaBox [0 0 595.2760 841.8900]"))
    }

    /** The whole point of the change: one file, every page inside it. */
    @Test
    fun `collects every page into one document`() {
        val bytes = write { document ->
            repeat(9) { index ->
                document.addPage { page -> page.text(10.0, 10.0, 8.0, "page $index") }
            }
        }
        val text = latin1(bytes)
        assertTrue(text.contains("/Type /Pages /Count 9"))
        assertEquals(9, Regex("/Type /Page /Parent").findAll(text).count())
        val kids = text.substringAfter("/Kids [").substringBefore("]").trim().split(" 0 R").size - 1
        assertEquals(9, kids)
    }

    /** A wrong xref table is the classic way a hand-rolled writer breaks silently. */
    @Test
    fun `every xref offset points at its own object header`() {
        val bytes = write { document ->
            repeat(3) { index ->
                document.addPage { page ->
                    page.drawJpeg(ByteArray(16) { it.toByte() }, 4, 4, 0.0, 0.0, 10.0, 10.0)
                    page.text(10.0, 10.0, 8.0, "page $index")
                }
            }
        }
        val text = latin1(bytes)
        // "startxref" also ends in "xref", so anchor on the newline before the table.
        val xrefAt = text.lastIndexOf("\nxref\n") + 1
        val startXref = text.substringAfterLast("startxref\n").substringBefore("\n").trim().toInt()
        assertEquals(xrefAt, startXref, "startxref does not point at the xref table")

        val lines = text.substring(xrefAt).lines()
        val count = lines[1].trim().split(" ")[1].toInt()
        // Two fonts, the tree, the catalogue, then image + content + dictionary per page,
        // plus the mandatory free entry at number 0.
        assertEquals(4 + 3 * 3 + 1, count, "unexpected object count")
        for (index in 1 until count) {
            val offset = lines[2 + index].substring(0, 10).toInt()
            assertTrue(
                text.startsWith("$index 0 obj", offset),
                "object $index: xref says $offset, found '${text.substring(offset, offset + 12)}'",
            )
        }
        assertTrue(text.contains("/Size $count"))
    }

    /** Pages name their parent before the tree exists, so the reserved number must hold. */
    @Test
    fun `pages point at the page tree that is written after them`() {
        val bytes = write { document ->
            repeat(2) { document.addPage { page -> page.text(0.0, 0.0, 8.0, "x") } }
        }
        val text = latin1(bytes)
        val parent = Regex("/Parent (\\d+) 0 R").findAll(text).map { it.groupValues[1] }.toSet()
        assertEquals(setOf("3"), parent)
        assertTrue(text.contains("3 0 obj\n<< /Type /Pages"))
        assertTrue(text.contains("/Root 4 0 R"))
    }

    @Test
    fun `jpeg payloads are embedded untouched as DCTDecode streams`() {
        val jpeg = ByteArray(64) { (it * 7).toByte() }
        val bytes = write { it.addPage { page -> page.drawJpeg(jpeg, 8, 8, 0.0, 0.0, 100.0, 100.0) } }
        val text = latin1(bytes)
        assertTrue(text.contains("/Subtype /Image /Width 8 /Height 8"))
        assertTrue(text.contains("/Filter /DCTDecode /Length 64"))
        assertTrue(latin1(jpeg) in text, "the jpeg bytes were altered")
        assertTrue(contentStreams(bytes).contains("100.0000 0 0 100.0000 0.0000 0.0000 cm /Im0 Do"))
    }

    @Test
    fun `parentheses and backslashes in labels are escaped`() {
        val bytes = write { it.addPage { page -> page.text(0.0, 0.0, 8.0, "a(b)c\\d") } }
        assertTrue(contentStreams(bytes).contains("(a\\(b\\)c\\\\d) Tj"))
    }

    /** WinAnsi, not Latin-1: the attribution line's em dash lives at 0x97. */
    @Test
    fun `typographic punctuation is transcoded to WinAnsi`() {
        val bytes = write {
            it.addPage { page -> page.text(0.0, 0.0, 8.0, "© IGN — SCAN25®") }
        }
        val expected = "(© IGN " + 0x97.toChar() + " SCAN25®) Tj"
        assertTrue(contentStreams(bytes).contains(expected), "em dash was not transcoded to WinAnsi")
    }

    @Test
    fun `content streams are deflated`() {
        val bytes = write { document ->
            document.addPage { page ->
                repeat(40) { page.text(10.0, it.toDouble(), 8.0, "un libelle assez repetitif") }
            }
        }
        val compressed = Regex("<< /Filter /FlateDecode /Length (\\d+) >>")
            .find(latin1(bytes))!!.groupValues[1].toInt()
        val plain = contentStreams(bytes).length
        assertTrue(compressed < plain / 2, "content stream barely shrank: $compressed vs $plain")
    }

    @Test
    fun `text width uses the real Helvetica metrics`() {
        var widths: PdfPage? = null
        write { it.addPage { page -> widths = page } }
        val page = widths!!
        // Every Helvetica digit is 556/1000 em.
        assertEquals(4 * 0.556 * 10.0, page.textWidth("1234", 10.0), 1e-9)
        assertTrue(page.textWidth("iii", 10.0) < page.textWidth("WWW", 10.0))
        assertTrue(page.textWidth("Page", 10.0, bold = true) > page.textWidth("Page", 10.0))
    }
}
